### ISSUE TITLE
Set lengths on indexed fields (for SQL Server, at least)

### ISSUE THREAD

**Issue by ChrisSimmons** [IdentityServer/IdentityServer4.EntityFramework#32]
Perhaps tangentially-related to #1, but the properties on which indexes are based should have reasonable max lengths.  For example, here is the output of the create script after migrating to current:

``` text
Warning! The maximum key length is 900 bytes. The index 'PK_PersistedGrants' has maximum length of 1800 bytes. For some combination of large values, the insert/update operation will fail.
Warning! The maximum key length is 900 bytes. The index 'IX_PersistedGrants_SubjectId_ClientId' has maximum length of 1300 bytes. For some combination of large values, the insert/update operation will fail.
Warning! The maximum key length is 900 bytes. The index 'IX_PersistedGrants_SubjectId_ClientId_Type' has maximum length of 2200 bytes. For some combination of large values, the insert/update operation will fail.
```

This is because `PersistedGrant.ClientId` is hard-limited to a width of 200 and `SubjectId`, `Key`, and `Type` are all limited to a length of 450 by EF Core itself (see [this](https://docs.efproject.net/en/latest/modeling/max-length.html#conventions)).

I know there was talk in #1 about putting limits like this in IdentityServer4 itself?

**Comment by scottbrady91** [comment]
Currently this project hasn't taken a hard dependency on a database provider, only on relational databases. Do you know if this is an issue on other database providers?

Also, did these warnings just appear when you ran `dotnet ef migrations add`? (So I can replicate).

**Comment by ChrisSimmons** [comment]
Understood on being DB provider-agnostic.  No idea on other databases.

On the migrations side, to replicate it would be:
* `dotnet ef migrations add Whatever` to get your migrations "to current"
* Either:
  * `dotnet ef migrations script LatestAppliedMigration -o out.sql` to script from an existing `LatestAppliedMigration` to current or ...
  * `dotnet ef migrations script -o out.sql` if you're starting from a clean DB
* Executing `out.sql` against the database.

Bottom line is if you have "reasonable" restrictions in whatever app/tool you use to maintain the client IDs and subject IDs, this really isn't an issue.  In my app, I plan to ensure that these are well below the sizes which would raise these DB failure.

IMHO, this is low priority but it's something that should be considered holistically with IdentityServer4.

**Comment by brockallen** [comment]
The insert will only fail if the combination exceeds that size right? IOW, you will find out at runtime for a particular combination that it's too damn long. Maybe that's acceptable behavior.

**Comment by brockallen** [comment]
Ok, I set Key to 200, Type to 50, Subject to 200, and ClientId was already 200. Sound legit?

**Comment by ChrisSimmons** [comment]
@brockallen, yes that's the described behavior.  You'll only see a failure when trying to INSERT or UPDATE an overlong value.  As for those limits, yes I believe that covers the problem.   Thank you!
